## Zeming Zhang
#### zemingzh@buffalo.edu
#### zemingzh

In [1]:
import csv
import pandas as pd
import numpy as np
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import MultiLabelBinarizer
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.neural_network import MLPClassifier

In [43]:
file_path = "title.basics.tsv"
comedy_romance_1994 = ('', float('inf'))

# Open the file using a context manager to ensure it's properly closed
with open(file_path, "r", encoding="utf-8") as f:

    # Use csv.reader to parse the file using the tab delimiter
    reader = csv.reader(f, delimiter='\t')

    # Iterate over each row in the file
    for row in reader:

        # Use a try-except block to handle rows that might not have the correct number of fields
        try:
            # Unpack the fields from the row
            tconst, title_type, primary_title, original_title, is_adult, start_year, end_year, runtime_minutes, genres = row

        except:
            # If there are not enough fields in the row, split the fields by tab and extend the row list
            second_split = []
            for i in row:
                second_split += i.split('\t')
            tconst, title_type, primary_title, original_title, is_adult, start_year, end_year, runtime_minutes, genres = second_split

        # Split the genres field into a list of individual genres
        genres_list = genres.split(',')

        # Check if the movie is a TV series, a comedy, a romance, and was released in 1994
        if title_type == 'tvSeries' and 'Comedy' in genres_list and 'Romance' in genres_list and start_year == '1994':
            
            # Convert the runtime to a float, or set it to infinity if it's missing or invalid
            try:
                runtime = float(runtime_minutes)
            except:
                print(f"{primary_title} has runtime of {runtime_minutes}")
                runtime = float('inf')

            # If the current TV series has a shorter runtime than the current shortest runtime, update the shortest runtime TV series
            if runtime < comedy_romance_1994[1]:
                comedy_romance_1994 = (primary_title, runtime)


Four Lives, Four Loves has runtime of \N
Tavros me toxoti has runtime of \N
I'll Make a Habit of It has runtime of \N


In [44]:
# Print the TV series with the shortest runtime
print(f"The TV series with the shortest runtime is '{comedy_romance_1994[0]}' with a runtime of {comedy_romance_1994[1]} minutes.")


The TV series with the shortest runtime is 'Friends' with a runtime of 22.0 minutes.


In [2]:
file_path = "title.basics.tsv"
rows = []

# Open the file using a context manager to ensure it's properly closed
with open(file_path, "r", encoding="utf-8") as f:

    # Use csv.reader to parse the file using the tab delimiter
    reader = csv.reader(f, delimiter='\t')

    # Iterate over each row in the file
    for row in reader:

        # Use a try-except block to handle rows that might not have the correct number of fields
        try:
            # Unpack the fields from the row
            tconst, title_type, primary_title, original_title, is_adult, start_year, end_year, runtime_minutes, genres = row

        except:
            # If there are not enough fields in the row, split the fields by tab and extend the row list
            second_split = []
            for i in row:
                second_split += i.split('\t')
            tconst, title_type, primary_title, original_title, is_adult, start_year, end_year, runtime_minutes, genres = second_split

        # Split the genres field into a list of individual genres
        genres_list = genres.split(',')
        
        # Append the row as a list to the rows list
        rows.append([tconst, title_type, primary_title, original_title, is_adult, start_year, end_year, runtime_minutes, genres_list])
        

In [3]:
# Create a pandas dataframe from the rows
df = pd.DataFrame(rows[1:], columns=['tconst', 'title_type', 'primary_title', 'original_title', 'is_adult', 'start_year', 'end_year', 'runtime_minutes', 'genres_list'])
df.head()

tconst title_type           primary_title          original_title  \
0  tt0000001      short              Carmencita              Carmencita   
1  tt0000002      short  Le clown et ses chiens  Le clown et ses chiens   
2  tt0000003      short          Pauvre Pierrot          Pauvre Pierrot   
3  tt0000004      short             Un bon bock             Un bon bock   
4  tt0000005      short        Blacksmith Scene        Blacksmith Scene   

  is_adult start_year end_year runtime_minutes                   genres_list  
0        0       1894       \N               1          [Documentary, Short]  
1        0       1892       \N               5            [Animation, Short]  
2        0       1892       \N               4  [Animation, Comedy, Romance]  
3        0       1892       \N              12            [Animation, Short]  
4        0       1893       \N               1               [Comedy, Short]

In [4]:
# This code drops four columns ('end_year', 'tconst', 'primary_title', and 'original_title') from the `df` dataframe
df.drop('end_year', axis=1, inplace=True)
df.drop('tconst', axis=1, inplace=True)
df.drop('primary_title', axis=1, inplace=True)
df.drop('original_title', axis=1, inplace=True)

# Display the first five rows of the modified dataframe
df.head()

title_type is_adult start_year runtime_minutes                   genres_list
0      short        0       1894               1          [Documentary, Short]
1      short        0       1892               5            [Animation, Short]
2      short        0       1892               4  [Animation, Comedy, Romance]
3      short        0       1892              12            [Animation, Short]
4      short        0       1893               1               [Comedy, Short]

In [5]:
# This code creates a new dataframe `cleaned_df` by dropping rows that contain '\\N' values in the 'runtime_minutes' column of the original `df` dataframe
cleaned_df = df.drop(df[df['runtime_minutes'] == '\\N'].index)

# Display the modified dataframe `cleaned_df`
cleaned_df

title_type is_adult start_year runtime_minutes  \
0            short        0       1894               1   
1            short        0       1892               5   
2            short        0       1892               4   
3            short        0       1892              12   
4            short        0       1893               1   
...            ...      ...        ...             ...   
9750363      movie        0       2013              49   
9750369  tvEpisode        0       2019              43   
9750404  tvEpisode        0       2014              11   
9750411      short        0       2015              27   
9750412  tvEpisode        0       2014              10   

                             genres_list  
0                   [Documentary, Short]  
1                     [Animation, Short]  
2           [Animation, Comedy, Romance]  
3                     [Animation, Short]  
4                        [Comedy, Short]  
...                                  ...  
9750363                    [Documentary]  
9750369  [Family, Game-Show, Reality-TV]  
9750404   [Adventure, Animation, Comedy]  
9750411                          [Short]  
9750412   [Adventure, Animation, Comedy]  

[2868627 rows x 5 columns]

In [6]:
# This code converts the 'runtime_minutes' column of the `cleaned_df` dataframe from object to float64 data type
cleaned_df['runtime_minutes'] = cleaned_df['runtime_minutes'].astype('float64')

# This code groups the `cleaned_df` dataframe by 'title_type' column and calculates the mean value of 'runtime_minutes' for each group
# The resulting values are rounded to the nearest integer using the `round()` method
cleaned_df.groupby('title_type')['runtime_minutes'].mean().round(0)

title_type
movie            90.0
short            13.0
tvEpisode        37.0
tvMiniSeries     90.0
tvMovie          72.0
tvSeries         46.0
tvShort          12.0
tvSpecial        89.0
video            67.0
videoGame       108.0
Name: runtime_minutes, dtype: float64

In [7]:
# This code calculates the rounded mean values of 'runtime_minutes' for each 'title_type' in the `cleaned_df` dataframe
mean_runtimes = cleaned_df.groupby('title_type')['runtime_minutes'].mean().round(0)

# This code converts the Pandas series `mean_runtimes` to a dictionary object `mean_dict`
mean_dict = mean_runtimes.to_dict()

# This code adds a key-value pair to `mean_dict` where the title type is 'tvPilot' and the mean runtime value is 0.0
mean_dict['tvPilot'] = 0.0

# This code displays the updated `mean_dict` dictionary object
mean_dict

{'movie': 90.0,
 'short': 13.0,
 'tvEpisode': 37.0,
 'tvMiniSeries': 90.0,
 'tvMovie': 72.0,
 'tvSeries': 46.0,
 'tvShort': 12.0,
 'tvSpecial': 89.0,
 'video': 67.0,
 'videoGame': 108.0,
 'tvPilot': 0.0}

In [8]:
# Replace the "\\N" values in the "runtime_minutes" column of df with the corresponding mean values
df['runtime_minutes'] = df.apply(lambda x: mean_dict[x['title_type']] if x['runtime_minutes'] == '\\N' else x['runtime_minutes'], axis=1)

# Show df
df

title_type is_adult start_year runtime_minutes  \
0            short        0       1894               1   
1            short        0       1892               5   
2            short        0       1892               4   
3            short        0       1892              12   
4            short        0       1893               1   
...            ...      ...        ...             ...   
9750408  tvEpisode        0       2010            37.0   
9750409  tvEpisode        0       2010            37.0   
9750410  tvEpisode        0       2010            37.0   
9750411      short        0       2015              27   
9750412  tvEpisode        0       2014              10   

                            genres_list  
0                  [Documentary, Short]  
1                    [Animation, Short]  
2          [Animation, Comedy, Romance]  
3                    [Animation, Short]  
4                       [Comedy, Short]  
...                                 ...  
9750408         [Action, Drama, Family]  
9750409         [Action, Drama, Family]  
9750410         [Action, Drama, Family]  
9750411                         [Short]  
9750412  [Adventure, Animation, Comedy]  

[9750413 rows x 5 columns]

In [10]:
# This code loops over all the columns in the DataFrame using the df.columns attribute, 
# and for each column, it drops all the rows where the cell value is equal to "\N".
for col in df.columns:
    df = df.drop(df[df[col] == '\\N'].index)
    
# Show df
df

title_type is_adult start_year runtime_minutes  \
0            short        0       1894               1   
1            short        0       1892               5   
2            short        0       1892               4   
3            short        0       1892              12   
4            short        0       1893               1   
...            ...      ...        ...             ...   
9750408  tvEpisode        0       2010            37.0   
9750409  tvEpisode        0       2010            37.0   
9750410  tvEpisode        0       2010            37.0   
9750411      short        0       2015              27   
9750412  tvEpisode        0       2014              10   

                            genres_list  
0                  [Documentary, Short]  
1                    [Animation, Short]  
2          [Animation, Comedy, Romance]  
3                    [Animation, Short]  
4                       [Comedy, Short]  
...                                 ...  
9750408         [Action, Drama, Family]  
9750409         [Action, Drama, Family]  
9750410         [Action, Drama, Family]  
9750411                         [Short]  
9750412  [Adventure, Animation, Comedy]  

[8432513 rows x 5 columns]

In [11]:
# This code converts the 'is_adult', 'start_year', and 'runtime_minutes' columns of the `df` dataframe to float64 data type
df['is_adult'] = df['is_adult'].astype('float64')
df['start_year'] = df['start_year'].astype('float64')
df['runtime_minutes'] = df['runtime_minutes'].astype('float64')

In [12]:
# Create a LabelEncoder object
le = LabelEncoder()

# Encode the title_type column
df['title_type'] = le.fit_transform(df['title_type'])

# Print the mapping between the original values and the encoded values
print(dict(zip(le.classes_, le.transform(le.classes_))))

{'movie': 0, 'short': 1, 'tvEpisode': 2, 'tvMiniSeries': 3, 'tvMovie': 4, 'tvPilot': 5, 'tvSeries': 6, 'tvShort': 7, 'tvSpecial': 8, 'video': 9, 'videoGame': 10}


In [13]:
# Show df
df.head()

title_type  is_adult  start_year  runtime_minutes  \
0           1       0.0      1894.0              1.0   
1           1       0.0      1892.0              5.0   
2           1       0.0      1892.0              4.0   
3           1       0.0      1892.0             12.0   
4           1       0.0      1893.0              1.0   

                    genres_list  
0          [Documentary, Short]  
1            [Animation, Short]  
2  [Animation, Comedy, Romance]  
3            [Animation, Short]  
4               [Comedy, Short]

In [14]:
# This code resets the index of the `df` dataframe and drops the old index
df = df.reset_index(drop=True)

# Show df
df

title_type  is_adult  start_year  runtime_minutes  \
0                 1       0.0      1894.0              1.0   
1                 1       0.0      1892.0              5.0   
2                 1       0.0      1892.0              4.0   
3                 1       0.0      1892.0             12.0   
4                 1       0.0      1893.0              1.0   
...             ...       ...         ...              ...   
8432508           2       0.0      2010.0             37.0   
8432509           2       0.0      2010.0             37.0   
8432510           2       0.0      2010.0             37.0   
8432511           1       0.0      2015.0             27.0   
8432512           2       0.0      2014.0             10.0   

                            genres_list  
0                  [Documentary, Short]  
1                    [Animation, Short]  
2          [Animation, Comedy, Romance]  
3                    [Animation, Short]  
4                       [Comedy, Short]  
...                                 ...  
8432508         [Action, Drama, Family]  
8432509         [Action, Drama, Family]  
8432510         [Action, Drama, Family]  
8432511                         [Short]  
8432512  [Adventure, Animation, Comedy]  

[8432513 rows x 5 columns]

In [15]:
# This code calculates the length of the 'start_year' column in the `df` dataframe
list_len= len(df['start_year'])

# Show len
list_len

8432513

In [16]:
# Create an instance of MultiLabelBinarizer
mlb = MultiLabelBinarizer()

# Fit the binarizer to the genres_list column and transform it into a binary matrix
mlb.fit_transform(df['genres_list'])

df_dict = {c: [0] * list_len for c in mlb.classes_}

idx = 0
for row in df['genres_list']:
    for i in row:
        df_dict[i][idx] = 1
    idx += 1

# Create a new DataFrame with the one-hot encoded genres
df_genres = pd.DataFrame(df_dict)

# Concatenate the new DataFrame with the original one
df = pd.concat([df, df_genres], axis=1)

# # Drop the original genres_list column
df = df.drop('genres_list', axis=1)

# Show df
df.head()

title_type  is_adult  start_year  runtime_minutes  Action  Adult  \
0           1       0.0      1894.0              1.0       0      0   
1           1       0.0      1892.0              5.0       0      0   
2           1       0.0      1892.0              4.0       0      0   
3           1       0.0      1892.0             12.0       0      0   
4           1       0.0      1893.0              1.0       0      0   

   Adventure  Animation  Biography  Comedy  ...  Reality-TV  Romance  Sci-Fi  \
0          0          0          0       0  ...           0        0       0   
1          0          1          0       0  ...           0        0       0   
2          0          1          0       1  ...           0        1       0   
3          0          1          0       0  ...           0        0       0   
4          0          0          0       1  ...           0        0       0   

   Short  Sport  Talk-Show  Thriller  War  Western  \N  
0      1      0          0         0    0        0   0  
1      1      0          0         0    0        0   0  
2      0      0          0         0    0        0   0  
3      1      0          0         0    0        0   0  
4      1      0          0         0    0        0   0  

[5 rows x 33 columns]

In [17]:
# This code drops all rows that contain '\\N' in the `df` dataframe
df = df.drop('\\N', axis=1)

# Show df
df.head()

title_type  is_adult  start_year  runtime_minutes  Action  Adult  \
0           1       0.0      1894.0              1.0       0      0   
1           1       0.0      1892.0              5.0       0      0   
2           1       0.0      1892.0              4.0       0      0   
3           1       0.0      1892.0             12.0       0      0   
4           1       0.0      1893.0              1.0       0      0   

   Adventure  Animation  Biography  Comedy  ...  News  Reality-TV  Romance  \
0          0          0          0       0  ...     0           0        0   
1          0          1          0       0  ...     0           0        0   
2          0          1          0       1  ...     0           0        1   
3          0          1          0       0  ...     0           0        0   
4          0          0          0       1  ...     0           0        0   

   Sci-Fi  Short  Sport  Talk-Show  Thriller  War  Western  
0       0      1      0          0         0    0        0  
1       0      1      0          0         0    0        0  
2       0      0      0          0         0    0        0  
3       0      1      0          0         0    0        0  
4       0      1      0          0         0    0        0  

[5 rows x 32 columns]

In [18]:
# This code splits the `df` dataframe into training and test sets
# X_train and X_test are the feature matrices for the training and test sets respectively
# y_train and y_test are the target vectors for the training and test sets respectively
# The test set size is 20% of the entire dataset
# The random state is set to 42 for reproducibility
X_train, X_test, y_train, y_test = train_test_split(df.drop(columns=['title_type']), df['title_type'], test_size=0.2, random_state=42)

In [19]:
# Train and test the Random Forest model
rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)
rf_pred = rf.predict(X_test)
rf_acc = accuracy_score(y_test, rf_pred)
print(f"Random Forest accuracy: {rf_acc}")

Random Forest accuracy: 0.9450270767380787


In [21]:
# Train and test the Decision Tree model
dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train, y_train)
dt_pred = dt.predict(X_test)
dt_acc = accuracy_score(y_test, dt_pred)
print(f"Decision Tree accuracy: {dt_acc}")

Decision Tree accuracy: 0.9430685862995797


## Result:
#### Random Forest accuracy: 0.9450270767380787 or 94.5027%
#### This is the best model out of all the models tested.